tests/test_set.py

```diff
@@ -131,6 +273,330 @@ def test_toml_metadata(get_mp3):
+    )
+
+    actual = subprocess.check_output(["id3manager", "-f", "toml", "get", test_mp3])
+    assert expected == actual.decode("utf-8")
+
+
+def test_toml_metadata_apic_file(get_mp3, testdata):
+    image_path = testdata / "logo.png"
+    image_data = open(image_path, "rb").read()
+
+    input = textwrap.dedent(
+        f"""\
+        [[TDRC]]
+        text = "2022-11-27"
+
+        [[TRCK]]
+        text = "14/14"
+
+        [[TPE1]]
+        text = "Ігор, Роман"
+
+        [[TALB]]
+        text = "Шо по коду?"
+
+        [[TIT2]]
+        text = "Обробка помилок"
+
+        [[TCON]]
+        text = "Podcast"
+
+        [[APIC]]
+        data = "file://{image_path}"
+
+        [[CHAP]]
+        text = "Кінець"
+        timestamp = "00:00:00"
+        """
+    )
+    expected = textwrap.dedent(
+        f"""\
+        [[TDRC]]
+        text = "2022-11-27"
+
+        [[TRCK]]
+        text = "14/14"
+
+        [[TPE1]]
+        text = "Ігор, Роман"
+
+        [[TALB]]
+        text = "Шо по коду?"
+
+        [[TIT2]]
+        text = "Обробка помилок"
+
+        [[TCON]]
+        text = "Podcast"
+
+        [[APIC]]
+        mime = "image/png"
+        data = "{base64.b64encode(image_data).decode()}"
+
+        [[CHAP]]
+        text = "Кінець"
+        timestamp = "00:00:00"
+        """
+    )
+
+    test_mp3 = get_mp3("metadata.mp3")
+    subprocess.check_output(
+        ["id3manager", "-f", "toml", "set", test_mp3], input=input.encode("utf-8")
+    )
+
+    actual = subprocess.check_output(["id3manager", "-f", "toml", "get", test_mp3])
+    assert expected == actual.decode("utf-8")
+
+
+def test_toml_metadata_apic_data(get_mp3):
+    expected = textwrap.dedent(
+        """\
+        [[TDRC]]
+        text = "2022-11-27"
+
+        [[TRCK]]
+        text = "14/14"
+
+        [[TPE1]]
+        text = "Ігор, Роман"
+
+        [[TALB]]
+        text = "Шо по коду?"
+
+        [[TIT2]]
+        text = "Обробка помилок"
+
+        [[TCON]]
+        text = "Podcast"
+
+        [[APIC]]
+        mime = "image/jpeg"
+        data = "ZHNhZmFzZmFkcw=="
+
+        [[CHAP]]
+        text = "Кінець"
+        timestamp = "00:00:00"
+        """
+    )
+    test_mp3 = get_mp3("metadata.mp3")
+    subprocess.check_output(
+        ["id3manager", "-f", "toml", "set", test_mp3], input=expected.encode("utf-8")
+    )
+
+    actual = subprocess.check_output(["id3manager", "-f", "toml", "get", test_mp3])
+    assert expected == actual.decode("utf-8")
+
+
+def test_toml_metadata_apic_data_url(get_mp3):
+    input = textwrap.dedent(
+        """\
+        [[TDRC]]
+        text = "2022-11-27"
+
+        [[TRCK]]
+        text = "14/14"
+
+        [[TPE1]]
+        text = "Ігор, Роман"
+
+        [[TALB]]
+        text = "Шо по коду?"
+
+        [[TIT2]]
+        text = "Обробка помилок"
+
+        [[TCON]]
+        text = "Podcast"
+
+        [[APIC]]
+        data = "data:image/jpeg;base64,ZHNhZmFzZmFkcw=="
```

Comment:
Hm.. It concerns me that we may end up in a situation with two mimetypes specified. E.g.

```
[[APIC]]
mime = image/png
data = "data:image/jpeg;base64,ZHNhZmFzZmFkcw=="
```

I feel that we should probably parse in unto three independent components:

* mimetype
* data
* cover_type

And then represent those components separately in TOML, and somehow as a single line in plain/text :thinking:

Reply:
Yeah, I can see how this can be confusing.

So the story here is that I started with the plain text format and I was looking for the most intuitive and/or "standard" way to configure this. I stumbled upon https://developer.mozilla.org/en-US/docs/Web/URI/Schemes/data#syntax, and that seemed like a great fit for embedding image data which had the look and feel of URLs and could easily be substituted with a normal `http://` or a `file://` URL to give people more options.

Now with TOML, we don't need to go through all that trouble, and we can simply use separate attributes. Perhaps, something like:

```toml
[[APIC]]
mime = image/png
data = ZHNhZmFzZmFkcw==
type = COVER_FRONT
```

```toml
[[APIC]]
mime = image/png
uri = file://path/to/a/local/file.png
type = COVER_FRONT
```

```toml
[[APIC]]
uri = https://path.to/a/remote/file.jpg
type = COVER_FRONT
```

with the restriction that you can only use one of `data` or `uri` at the same time, and the former always expects a Base64 encoded payload and requires `mime` to be specified, while the URI version only requires the location. I didn't go for this option because I wanted to make the TOML configuration follow the plain text configuration as close as possible, but, perhaps, that creates more issues than it solves...

WDYT?

Reply:
@malor wrote:
> and the former always expects a Base64 encoded payload and requires mime to be specified

Is mimetype required in ID3 APIC frame? The [ID3v2 spec](https://id3.org/id3v2.3.0#Attached_picture) sounds unclear to me:

> _In the event that the MIME media type name is omitted, "image/" will be implied._ 

What does `image/` mean? Is this some kind of generic mimetype for _any_ image type? Or do they imply that some _random_ image mimetype should be used instead?

It's important for me to understand in order to make a concisious decision on how to proceed with mime type.

Oh.. are you suggesting `mime` type to be _required_ in order to render it in plain text format? According to [Data URLs](https://developer.mozilla.org/en-US/docs/Web/URI/Schemes/data#syntax), the mime type there is optional.

Reply:
> What does image/ mean? Is this some kind of generic mimetype for any image type? Or do they imply that some random image mimetype should be used instead?

https://developer.mozilla.org/en-US/docs/Web/HTTP/MIME_types claims that:

> Each type has its own set of possible subtypes. A MIME type **always** has both a type and a subtype, never just one or the other.

So, I think, `image/` is useless because it essentially means that the consumer either need to guess the correct format or just try all supported formats one by one until it finds the implementation that can successfully interpret the payload.

> Oh.. are you suggesting mime type to be required in order to render it in plain text format? According to [Data URLs](https://developer.mozilla.org/en-US/docs/Web/URI/Schemes/data#syntax), the mime type there is optional.

Yes. My suggestion would be to make it required as we need to put a valid MIME type value in the APIC frame to make it usable for players. I tried _not_ setting the MIME type, and, as expected, `mpv` refuses to display the cover image:

```shell
> mpv output.mp3
[ffmpeg] Unknown attached picture mimetype: , skipping.
● Audio  --aid=1  (mp3 1ch 48000 Hz 115 kbps)
```

We can make it required, and then use data URLs in plain text (that combine the MIME type and the payload), and separate attributes in TOML.

In the plain text implementation, we would still need a way to specify the image type somehow, so my suggestion from the comment thread above is to supplement the data URL with an optional image type value (either the string name, or the hexadecimal number from https://id3.org/id3v2.3.0#Attached_picture) separated by a space.

Reply:
> We can make it required, and then use data URLs in plain text (that combine the MIME type and the payload), and separate attributes in TOML.

While I agree that it's useless to have APIC w/o mimetype, I still don't understand why to enforce it. I mean, the data URLs do not require a mimetype. It's optional there:

```
data:;base64,ZHNhZmFzZmFkcw==
```

So I'm trying to understand why to enforce that? 

What if are to use this project with a media file that misses mime type? It'd be unfortunate if you won't be able to set() what you previously get().